### ISSUE TITLE
virtualization

### ISSUE THREAD

**Issue by caridy** [BuilderIO/partytown#15]
hey @adamdbradley, thanks for putting effort on this, this is a great area of exploration. there are few things that I will like to suggest or hint based on the little that I know so far about this project (which isn't much, I haven't get a chance to read the whole code yet).

based on the goals (from the readme), what you're trying to achieve is basically a virtualization mechanism, specifically, a virtual environment for the DOM, and from that perspective, there are few things that I will suggest to consider:

1. the object graph that is reachable from the DOM is vast, but the entry points to it are not so much, in fact, it is less than 1k in property descriptors (basically, language + global names in Web IDL). If you get to virtualize those, then any program running in the virtual environment must go thru them to access the DOM.
2. reconstructing the DOM APIs is a tricky business, many have tried, it is privative if you try to boot it up, it is also always falling behind, it is a catch up game if you try to replicate the APIs. This is where a membrane comes handy, because you can basically reconstruct the DOM APIs in a lazy manner (on first access), and will only require to virtualize the global names first. Keep in mind that a complex program only really touches a tiny fragment of the vast Web IDL.
3. I see many similarities between the work you're doing here and the ShadowRealm (former Realm) TC39 Proposal that we have been pushing for quite some time as a new addition to the language. Obviously the biggest difference is that a ShadowRealm runs in the same process, but the rest is very similar, in principle. A ShadowRealm does not allow connection with the object graph of the incubator realm, just like the worker doesn't. You can have actionable integration between the two realms (in the case of the ShadowRealm via wrapped callables, in the case of the worker, via the sync XHR). The beauty here is that, in theory, you can build wrapped callables on top of the sync XHR protocol, and here is where things becomes a lot more interesting.
4. The DOM is very tricky to virtualize in part because of the many "sinks" that turn into evaluations, a good example of it is what to do when the virtualize script inserts another script tag. The way we have been thinking about this is by providing what we call "distortions", which will allow to control the behavior in a way that is not observable by the virtualized code. Membranes are pretty good at this.
5. Interaction with multiple window objects is also very tricky, imagine a script virtualized that creates an iframe, and attempt to interact with the `contentWindow` of that iframe. That gets very tricky because the identity discontinuity nature of that interaction, you can't use the worker's realm for the iframe, because that's not the expectation of the script creating the iframe.

These are just some of the ideas and challenges that we have encountered along the way, we can chat more about it.

**Comment by adamdbradley** [comment]
Yeah I've certainly hit many of the areas you're talking about. The sync communication is a challenge, but really I've found the real challenge is the cross-communcation between window contexts. I have been adding many tests covering these, mainly in the iframe tests page: https://github.com/BuilderIO/partytown/blob/main/tests/iframe/index.html

As for reconstructing the dom, there's very little that's being re-implemented, but instead we're heavily using JS Proxies, and the calls are being forwarded through the sync communication, and same call is applied to the main thread, then return values come back through. Any specific DOM implementations I've added have been more of an optimization to reduce calls to main.

Another example is that neither `addEventListener()` and `removeEventListener()` is implemented, but when the proxies receives the method call, that call is send the main thread, and the arguments are serialized. And in this case, the serialized callback function is persisted in the same thread context and given and unique ID. And then when that ID is called on the main thread, it knows to call the function on the worker thread.

I'd love to hear more about ShadowRealm. Where would I find a good resource on that? Thanks!

**Comment by caridy** [comment]
> I have been adding many tests covering these, mainly in the iframe tests page: https://github.com/BuilderIO/partytown/blob/main/tests/iframe/index.html

I will take a look this part is certainly very tricky.

> As for reconstructing the dom, there's very little that's being re-implemented, but instead we're heavily using JS Proxies, and the calls are being forwarded through the sync communication, and same call is applied to the main thread, then return values come back through. Any specific DOM implementations I've added have been more of an optimization to reduce calls to main.

This must be a membrane, which is probably what you have already even if not formalized. Membranes are hard, and very tricky. At SES and TC39 we use the term "near-membrane" for this kind of membrane that give the program the illusion that they are running in the "blue" side when in reality they are running in the "red" side. On top of that, think about a library that is attempting to patch `Array.prototype.foo`, should that ended up being a modification on the main window, or should that be observed only in the worker? What about a DOM API? `HTMLElement.prototype.foo`? Those are good questions in terms of goals of this library. A good example here is an older google map distribution that patches and "breaks" some language features. A more simple example of this sort of mechanism is a conflict between libraries that can be resolved by running them in separate workers but still giving them the illusion that they are running in the main window. The good news here is that we have some good precedents about this that you might be able to reuse.

> Another example is that neither addEventListener() and removeEventListener() is implemented, but when the proxies receives the method call, that call is send the main thread, and the arguments are serialized. And in this case, the serialized callback function is persisted in the same thread context and given and unique ID. And then when that ID is called on the main thread, it knows to call the function on the worker thread.

Again, implementing a tailor made proxy layer for this is going to be difficult, proxies are hard, and using IDs is also problematic because of potential leakages, and such. A membrane implementation might help. Here is a suggestion of the one we use in production at salesforce: https://github.com/salesforce/near-membrane/blob/master/packages/near-membrane-base/src/membrane.ts (beware, this was designed to work with ShadowRealms - when it lands in browsers - but also work fine today in a detached iframe, and I will also work for a worker if you figure how to implement the underlaying layer over sync XHR).

> I'd love to hear more about ShadowRealm. Where would I find a good resource on that? Thanks!

https://github.com/tc39/proposal-shadowrealm/
